networkx/algorithms/lowest_common_ancestors.py

```diff
@@ -10,9 +10,121 @@
     "all_pairs_lowest_common_ancestor",
     "tree_all_pairs_lowest_common_ancestor",
     "lowest_common_ancestor",
+    "naive_lowest_common_ancestor",
+    "naive_all_pairs_lowest_common_ancestor",
 ]
 
 
+def _get_a_lowest_common_ancestor(G, common_ancestors):
+    """Return a lowest common ancestor from the given set of common ancestors"""
+    # Start with an arbitrary node v from the set of common ancestors.  Follow
+    # arbitrary outgoing edges, remaining in the set of common ancestors, until
+    # reaching a node with no outgoing edge to another of the common ancestors.
+
+    v = next(iter(common_ancestors))
+    while True:
+        successor = None
+        for w in G.successors(v):
+            if w in common_ancestors:
+                successor = w
+                break
+        if successor is None:
+            return v
+        v = successor
+
+
+@not_implemented_for("undirected")
+@not_implemented_for("multigraph")
+def naive_all_pairs_lowest_common_ancestor(G, pairs=None):
+    """Return the lowest common ancestor of all pairs or the provided pairs
+
+    Parameters
+    ----------
+    G : NetworkX directed graph
+
+    pairs : iterable of pairs of nodes, optional (default: all pairs)
+        The pairs of nodes of interest.
+        If None, will find the LCA of all pairs of nodes.
+
+    Returns
+    -------
+    An iterator over ((node1, node2), lca) where (node1, node2) are
+    the pairs specified and lca is a lowest common ancestor of the pair.
+    Note that for the default of all pairs in G, we consider
+    unordered pairs, e.g. you will not get both (b, a) and (a, b).
+
+    Notes
+    -----
+    Only defined on non-null directed acyclic graphs.
+
+    See Also
+    --------
+    naive_lowest_common_ancestor
+    """
+    if not nx.is_directed_acyclic_graph(G):
+        raise nx.NetworkXError("LCA only defined on directed acyclic graphs.")
+    elif len(G) == 0:
+        raise nx.NetworkXPointlessConcept("LCA meaningless on null graphs.")
+    elif None in G:
+        raise nx.NetworkXError("None is not a valid node.")
+
+    ancestor_cache = {}
+
+    if pairs is None:
+
+        pairs = combinations_with_replacement(G, 2)
+
+    for v, w in pairs:
+        if v not in ancestor_cache:
+            ancestor_cache[v] = nx.ancestors(G, v)
+            ancestor_cache[v].add(v)
+        if w not in ancestor_cache:
+            ancestor_cache[w] = nx.ancestors(G, w)
+            ancestor_cache[w].add(w)
+
+        common_ancestors = ancestor_cache[v] & ancestor_cache[w]
+        if common_ancestors:
+            yield ((v, w), _get_a_lowest_common_ancestor(G, common_ancestors))
```

Comment:
Is this the only place this function is called?  If so, you might consider in-lining it. I'm fine whether you do or not...